src/downloader/downloader.cpp

```diff
@@ -0,0 +1,182 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <thread>
+
+#include <Poco/FileStream.h>
+#include <Poco/StreamCopier.h>
+#include <Poco/Task.h>
+#include <Poco/ThreadPool.h>
+#include <curl/curl.h>
+
+#include <utils/exception.hpp>
+
+#include "downloader.hpp"
+#include "logger/logmodule.hpp"
+
+namespace aos::mp::downloader {
+
+/***********************************************************************************************************************
+ * Static
+ **********************************************************************************************************************/
+
+static size_t WriteData(void* ptr, size_t size, size_t nmemb, FILE* stream)
+{
+    size_t written = fwrite(ptr, size, nmemb, stream);
+
+    return written;
+}
+
+static std::string GetFileNameFromURL(const std::string& url)
+{
+    Poco::URI   uri(url);
+    std::string path = uri.getPath();
+
+    if (uri.getScheme() == "file") {
+        if (path.empty() && !uri.getHost().empty()) {
+            path = "/" + uri.getHost();
+        }
+    }
+
+    size_t pos = path.find_last_of('/');
+    if (pos != std::string::npos) {
+        return path.substr(pos + 1);
+    }
+
+    return path;
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Downloader::Downloader(const std::string& downloadDir)
+    : mDownloadDir(downloadDir)
+    , mTaskManager(Poco::ThreadPool::defaultPool())
+{
+    try {
+        if (!std::filesystem::exists(mDownloadDir)) {
+            std::filesystem::create_directories(mDownloadDir);
+        }
+    } catch (const std::exception& e) {
+        AOS_ERROR_THROW("Failed to create download directory: downloadDir=" + mDownloadDir, aos::ErrorEnum::eFailed);
+    }
+}
+
+aos::RetWithError<std::string> Downloader::Download(const std::string& url)
+{
+    LOG_DBG() << "Sync downloading: url=" << url.c_str();
+
+    std::string outfilename = std::filesystem::path(mDownloadDir).append(GetFileNameFromURL(url)).string();
+
+    return {outfilename, RetryDownload(url, outfilename)};
+}
+
+Downloader::~Downloader()
+{
+    mTaskManager.cancelAll();
+    mTaskManager.joinAll();
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+aos::Error Downloader::Download(const std::string& url, const std::string& outfilename)
+{
+    Poco::URI uri(url);
+    if (uri.getScheme() == "file") {
+        return DownloadFile(uri, outfilename);
+    }
+
+    std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> curl(curl_easy_init(), curl_easy_cleanup);
```

Comment:
consider calling curl_global_init in the app's main function.

Reply:
I think it's better to keep curl_global_init in this function because it makes it clearer which module is using the curl functionality. Additionally, since this function isn't called frequently for the download service, I believe it's better to allocate and free resources as needed rather than keeping them allocated for the entire program's lifetime.

Reply:
If you did not already call [curl_global_init](https://curl.se/libcurl/c/curl_global_init.html) before calling this function, [curl_easy_init](https://curl.se/libcurl/c/curl_easy_init.html) does it automatically. This can be lethal in multi-threaded cases for platforms where [curl_global_init](https://curl.se/libcurl/c/curl_global_init.html) is not thread-safe, and it may then result in resource problems because there is no corresponding cleanup.

Unfortunately I don't know if curl_global_init is thread safe on our platforms.
So I believe it's easier to explicitly call it once somewhere in our app

Reply:
Done. Moved to where the app init